### DISCUSSION TITLE
How can I construct the Misorientation Axis in the IPF using GROD for both high and low GOS values?

### DISCUSSION THREAD

**Discussion by rockdef** [mtex-toolbox/mtex#1943] [Ask Anything] [answered]
Dear MTEX community, 

1) I want to plot the misorientation axis (MOA) for both high and low GOS values using Grain Reference Orientation Deviation (GROD). I'm encountering some errors. I would be grateful if someone could help or share their code. I have attached my versions below: 
```matlab
% GROD Calculations
ebsd_hbl = ebsd('Hornblende');
[grains_hbl,ebsd_hbl.grainId] = calcGrains(ebsd_hbl('Hornblende'));

% Remove one-three pixel grains
ebsd_hbl(grains_hbl(grains_hbl.grainSize <= 3)) = [];
[grains_hbl,ebsd_hbl.grainId] = calcGrains(ebsd_hbl('Hornblende'),'threshold',[2*degree, 10*degree]);

% Smooth grain boundaries
grains_hbl = smooth(grains_hbl,5);

% Denoise the orientations
F = halfQuadraticFilter;
ebsd_hbl = smooth(ebsd_hbl,F,grains_hbl,'fill');


% GOS threshold
grains_HGOS = grains((grains('Hornblende').GOS./degree>5));
grains_LGOS = grains((grains('Hornblende').GOS./degree<5));

% take the avarage of the misorientation angles for each grain
grod = ebsd.calcGROD(grains_HGOS);


% MOA in Crystal Coordinates/reference frame (CC)
figure
axCrystal = grod.axis;
plot(axCrystal,'contourf','fundamentalRegion','antipodal','figSize','large')
DataNum = numel(axCrystal.h);
mtexTitle(sprintf('Hbl %d data points', DataNum))
mtexColorbar('title','M.U.D.');

% MOA in Specimen Coordinates/reference frame (SC)
figure
axSpecimen = ebsd_hbl.orientations .* grod.axis('noSymmetry');
plot(axSpecimen,'contourf','fundamentalRegion','halfwidth',10*degree,'hemisphere','upper')%,'colorrange', [0 4])%
DataNum = numel(axSpecimen.x);
mtexTitle(sprintf('Hbl %d data points', DataNum))
mtexColorbar('title','M.U.D.');
```

2) How do I colorize two groups of GOS values (high and low) with different colors in the histogram plot?
```matlab
%% GOS vs. frequency (%)
figure 
histogram(grains.GOS./degree,60); 
xlabel('GOS (degree)');
ylabel('Frequency(%)');
```

I would really appreciate it if someone help me with this. Thank you very much in advance!

**Comment by ralfHielscher** [comment, marked as answer]
You can do something like
```matlab
highGOS = grod.angle > 5*degree;
lowGOS =  grod.angle < 5*degree;
% plot only highGOS
plot(axCrystal(highGOS),'contourf','fundamentalRegion','antipodal','figSize','large')
```
I hope this helps,
Ralf.

**Reply by rockdef** [reply]
Hi Ralf,

It works for me. Thank you very much for the help.

**Comment by rockdef** [comment]
Hi Ralf, 

While plotting, I encountered another issue. 

What is the command to get the number of data points or pixels and one point per grain for high and low GOS values? 
Looking forward to hearing from you. Thank you!

**Reply by ralfHielscher** [reply]
Could you please formulate you question more clearly?
The number of data points you can get by the command `length`, e.g. `length(ebsd)`
If you want to compute `grod` for grains, have a look at the command [`grainMean`](https://mtex-toolbox.github.io/EBSD.grainMean.html)

**Reply by rockdef** [reply]
Hi Ralf, 
I'm sorry for the delay because of unavoidable situations. 

I wanted to compute grod for grains with highGOS (>3*degree). Unfortunately, I won't be able to do that. Below is the code. 

% Threshold condition 
highGOS = grod.angle > 3*degree;

% Plot only highGOS
figure
plot(axCrystal(highGOS),'contourf','fundamentalRegion','antipodal','figSize','large')

DataNum = numel(axCrystal.h);
mtexTitle(sprintf('cal %d data points', DataNum))
mtexColorbar('title','M.U.D.');
mtexTitle ('High GOS ($>$3*degree)')
hold on
annotate([Miller({0,0,0,1},{-1,2,-1,0},{0,1,-1,2}, {1,0,-1,4},ebsd('Calcite').CS)],'label',{'C[0001]','a[-12-10]','f[01-12]','r[10-14]'},'MarkerSize',15,'FontSize',13, 'BackgroundColor','w');
hold off

Many thanks for considering my request!

**Reply by zmichels** [reply]
You shared some code to try and plot something... and you mentioned you are unable to do it... can you share some more information about what result you get (error?, or plot looks different than expected?) versus what result you expected?

Cheers,
Zach

**Comment by rockdef** [comment]
Hi Zach, 

1) Misorientation Axis (MOA) in Crystal Coordinates

grains = grains ('Calcite');
% compute the grain reference orientation deviation
grod = ebsd.calcGROD(grains);
axCrystal = grod.axis;
plot(axCrystal,'contourf','fundamentalRegion','antipodal','figSize','small')
mtexColorbar('title','distribution of misorientation axes in mrd')

2) Continuing from the previous code, my aim to create plots illustrating the Inverse Pole Figure (IPF) for GROD in both highGOS and lowGOS scenarios. 
To achieve this, I command the code: 

% Threshold condition 
highGOS = grod.angle > 3*degree;
% compute the grain reference orientation deviation
grod = ebsd.calcGROD(highGOS);
% Plot only highGOS
figure
plot(axCrystal(highGOS),'contourf','fundamentalRegion','antipodal','figSize','large')
mtexColorbar('title','M.U.D.');

Here's what I'm interested in plotting: I want to visualize the number of grains that have contributed to the highGOS and lowGOS IPF, along with the corresponding number of pixels for each IPF.

However, I've encountered an error in the process: "Index exceeds the number of array elements. The index must not exceed 434."

I can't track where are the errors and how to solve them. But I believe there will be another simpler command than this. 
Thank you very much for the help and time. Looking forward to hearing from you!

**Reply by zmichels** [reply]
In the error in the command screen it should tell you the line and function where the error is emanating. For example I get an error just trying to use the calcGROD function on any subset of grains... like it works on the initial grainset with the EBSD data... but as soon as I subset the grains it gives an error. I tried subsetting the associated EBSD, too... but it continues to error out. It would be helpful to know if your errors are related to the same problem (a potential issue/hurdle with using calcGROD).

**Reply by rockdef** [reply]
Hi Zach, 

Thank you very much for the help!

**Comment by zmichels** [comment]
%% 
mtexdata forsterite


phase = 'forsterite';

%% construct grain boundaries
[grains,ebsd.grainId,ebsd.mis2mean]  = calcGrains(ebsd('indexed'),'threshold',[1 10]*degree,'boundary','tight');
grains = grains(grains.grainSize>2);
ebsd = ebsd(grains);
[grains,ebsd.grainId,ebsd.mis2mean]  = calcGrains(ebsd('indexed'),'threshold',[1 10]*degree,'boundary','tight');
grains = smooth(grains,5);

%% GROD
% assigin GROD rotations as a property to EBSD variable, so can be subset
% later by grains

ebsd.prop.GROD = calcGROD(ebsd,grains);

%% GOS separation

GOSthresh = 5*degree;

highGOSgrains = grains(grains.GOS>GOSthresh);
lowGOSgrains = grains(grains.GOS<=GOSthresh);

highGOSebsd = ebsd(highGOSgrains);
lowGOSebsd = ebsd(lowGOSgrains);


%% GROD axes for HIGH-GOS grains
figure,
plot(highGOSebsd.GROD.axis,'antipodal','lower','smooth','fundamentalRegion')
cb = mtexColorbar('Title','M.U.D.');
annotation('textbox',[0 .7 0 .3],'String',sprintf('axes = %i\ngrains = %i',length(highGOSebsd),length(unique(highGOSebsd.grainId))),'FitBoxToText','on','LineStyle','none');


%% GROD axes for LOW-GOS grains
figure,
plot(lowGOSebsd.GROD.axis,'antipodal','lower','smooth','fundamentalRegion')
cb = mtexColorbar('Title','M.U.D.');
annotation('textbox',[0 .7 0 .3],'String',sprintf('axes = %i\ngrains = %i',length(lowGOSebsd),length(unique(lowGOSebsd.grainId))),'FitBoxToText','on','LineStyle','none');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% NOTE...
% unless specifying a reference orientation in each grain that is
% *different* from the mean orientation... it might be easier to use the
% "mis2mean" property. Here's an example:


%% mis2mean axes for HIGH-GOS grains
figure,
ax = ebsd(grains(grains.GOS>GOSthresh)).mis2mean.axis;
plot(ax,'antipodal','lower','smooth','fundamentalRegion')
cb = mtexColorbar('Title','M.U.D.');
annotation('textbox',[0 .7 0 .3],'String',sprintf('axes = %i\ngrains = %i',length(ax),sum(grains.GOS>GOSthresh)),'FitBoxToText','on','LineStyle','none');

**Comment by zmichels** [comment]
This works nicely for "mis2mean" GROD of a single phase...

`
%% mis2mean axes for HIGH-GOS grains
figure,
axHigh = ebsd(phase,grains(grains.GOS>GOSthresh)).mis2mean;
plotAxisDistribution(axHigh,'antipodal','lower','smooth','fundamentalRegion')
cb = mtexColorbar('Title','M.U.D.');
mtexTitle('GROD axes (High GOS)')
annotation('textbox',[.5 .6 0 .3],'String',sprintf('axes = %i\ngrains = %i',length(axHigh),sum(grains.GOS>GOSthresh)),'FitBoxToText','on','LineStyle','none');
hgexport(gcf,'highGOS_mis2mean_GROD_axes_xtal.png', hgexport('factorystyle'), 'Format', 'png')

%% mis2mean axes for LOW-GOS grains
figure,
axLow = ebsd(phase,grains(grains.GOS<=GOSthresh)).mis2mean;
plotAxisDistribution(axLow,'antipodal','lower','smooth','fundamentalRegion')
cb = mtexColorbar('Title','M.U.D.');
mtexTitle('GROD axes (Low GOS)')
annotation('textbox',[.5 .6 0 .3],'String',sprintf('axes = %i\ngrains = %i',length(axLow),sum(grains.GOS<=GOSthresh)),'FitBoxToText','on','LineStyle','none');
hgexport(gcf,'lowGOS_mis2mean_GROD_axes_xtal.png', hgexport('factorystyle'), 'Format', 'png')


`

![highGOS_mis2mean_GROD_axes_xtal](https://github.com/mtex-toolbox/mtex/assets/8769619/c1d5841e-0e93-4a15-a0a6-eef2a4b21403)
![lowGOS_mis2mean_GROD_axes_xtal](https://github.com/mtex-toolbox/mtex/assets/8769619/2f6a2c19-7d74-4d2b-9f39-44f17520492e)

**Reply by rockdef** [reply]
Hi Zach, 

It's perfectly working for me. I understood more clearly than before. 
Thank you very much for the help.
